pyg_lib/csrc/random/cpu/randint_engine.h

```diff
@@ -0,0 +1,88 @@
+#pragma once
+
+#include <torch/torch.h>
+
+#include <limits.h>
+
+namespace pyg {
+namespace random {
+
+const int RAND_PREFETCH_THRESHOLD = 128;
+
+// Use torch::randint to generate 64-bit random numbers
+const int RAND_PREFETCH_BITS = 64;
+
+class PrefetchedRandint {
+ public:
+  PrefetchedRandint(int size = RAND_PREFETCH_THRESHOLD,
+                    int bits = RAND_PREFETCH_BITS) {
+    prefetch_();
+  }
+
+  template <typename T>
+  T next(T range) {
+    unsigned needed = 64;
+
+    // Mutiple levels of range to save prefetched bits
+    if (range <= (1 << 15)) {
+      needed = 16;
+    } else if (range <= (1UL << 31)) {
+      needed = 32;
+    }
+
+    if (bits_ < needed) {
+      if (size_ > 0) {
+        size_--;
+        bits_ = RAND_PREFETCH_BITS;
+      } else {
+        // Prefetch if no enough bits
+        prefetch_();
+      }
+    }
+
+    // Currently torch could only make 64-bit signed numbers
+    uint64_t* prefetch_ptr =
+        reinterpret_cast<uint64_t*>(prefetched_randint_.data_ptr<int64_t>());
+    uint64_t mask = (needed == 64) ? std::numeric_limits<uint64_t>::max()
+                                   : (1ULL << needed) - 1;
+    uint64_t res = (prefetch_ptr[size_ - 1] & mask) % range;
```

Comment:
I feel if `needed=16`, and `bits_=64`, the next 4 calls of `rand(1<<15)` will return the same number because `size_` is the same, `mask` is the same, and `range` is the same.

Reply:
The prefetched element is shifted accordingly after that.